ipn/local.go

```diff
@@ -184,6 +179,118 @@ func (b *LocalBackend) SetDecompressor(fn func() (controlclient.Decompressor, er
 	b.newDecompressor = fn
 }
 
+func (b *LocalBackend) setClientStatus(newSt controlclient.Status) {
+	if newSt.LoginFinished != nil {
+		// Auth completed, unblock the engine
+		b.blockEngineUpdates(false)
+		b.authReconfig()
+		b.send(Notify{LoginFinished: &empty.Message{}})
+	}
+	if newSt.Persist != nil {
+		persist := *newSt.Persist // copy
+
+		b.mu.Lock()
+		b.prefs.Persist = &persist
+		prefs := b.prefs.Clone()
+		stateKey := b.stateKey
+		b.mu.Unlock()
+
+		if stateKey != "" {
+			if err := b.store.WriteState(stateKey, prefs.ToBytes()); err != nil {
+				b.logf("Failed to save new controlclient state: %v", err)
+			}
+		}
+		b.send(Notify{Prefs: prefs})
+	}
+	if newSt.NetMap != nil {
+		// Netmap is unchanged only when the diff is empty.
+		changed := true
+		b.mu.Lock()
+		if b.netMapCache != nil {
+			diff := newSt.NetMap.ConciseDiffFrom(b.netMapCache)
+			if strings.TrimSpace(diff) == "" {
+				changed = false
+				b.logf("netmap diff: (none)")
+			} else {
+				b.logf("netmap diff:\n%v", diff)
+			}
+		}
+		disableDERP := b.prefs != nil && b.prefs.DisableDERP
+		b.netMapCache = newSt.NetMap
+		b.mu.Unlock()
+
+		b.send(Notify{NetMap: newSt.NetMap})
+		// There is nothing to update if the map hasn't changed.
+		if changed {
+			b.updateFilter(newSt.NetMap)
+			b.updateDNSMap(newSt.NetMap)
+		}
+		if disableDERP {
+			b.e.SetDERPMap(nil)
+		} else {
+			b.e.SetDERPMap(newSt.NetMap.DERPMap)
+		}
+	}
+	if newSt.URL != "" {
+		b.logf("Received auth URL: %.20v...", newSt.URL)
+
+		b.mu.Lock()
+		interact := b.interact
+		b.authURL = newSt.URL
+		b.mu.Unlock()
+
+		if interact > 0 {
+			b.popBrowserAuthNow()
+		}
+	}
+	if newSt.Err != "" {
+		// TODO(crawshaw): display in the UI.
+		b.logf("Received error: %v", newSt.Err)
+		return
+	}
+	if newSt.NetMap != nil {
+		b.mu.Lock()
+		if b.state == NeedsLogin {
+			b.prefs.WantRunning = true
+		}
+		prefs := b.prefs
+		b.mu.Unlock()
+
+		b.SetPrefs(prefs)
+	}
+	b.stateMachine()
+}
+
+func (b *LocalBackend) setWgengineStatus(s *wgengine.Status, err error) {
+	if err != nil {
+		b.logf("wgengine status error: %#v", err)
+		return
+	}
+	if s == nil {
+		b.logf("weird: non-error wgengine update with status=nil: %v", s)
```

Comment:
we've been using `[unexpected]` as the convention for things that shouldn't fire (instead of "weird" here)